core/includes/file.inc

```diff
@@ -2018,6 +2021,74 @@ function file_validate_image_orientation(File $file, $orientation = 0) {
   return $errors;
 }
 
+/**
+ * Validate uploaded SVG files.
+ *
+ * @param File $file
+ *   A file entity.
+ *
+ * @return array
+ *   If the file is an SVG and is either not valid or contains
+ *   dangerous content, the array will contain an error message.
+ *
+ * @see hook_file_validate()
+ */
+function file_validate_svg(File $file) {
+  $errors = array();
+
+  if ($file->filemime == 'image/svg+xml') {
+    $file_contents = file_get_contents($file->uri);
+    libxml_use_internal_errors(TRUE);
+    $xml = simplexml_load_string($file_contents);
+    $errors = libxml_get_errors();
+
+    if (!$xml || !empty($errors)) {
+      return array(t('Invalid SVG file.'));
+    }
+    else {
+      $errors = array();
+      $namespaces = $xml->getNamespaces();
+      if (!in_array('http://www.w3.org/2000/svg', $namespaces)) {
+        $errors[] = t('Invalid SVG namespace.');
+      }
+
+      $search_patterns = array(
+        // @see https://developer.mozilla.org/en-US/docs/Web/SVG/Element
+        '//svg:foreignObject',
+        '//svg:script',
+        '//html:iframe',
+        // @see https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute
+        '//@*[starts-with(name(), "on")]',
+        '//svg:a[starts-with(@href, "javascript:")]',
+        '//svg:a[starts-with(@xlink:href, "javascript:")]',
+        '//svg:a[starts-with(@href, "data:")]',
+        '//svg:a[starts-with(@xlink:href, "data:")]',
+      );
+
+      $xml->registerXPathNamespace('svg', 'http://www.w3.org/2000/svg');
+      $xml->registerXPathNamespace('html', 'http://www.w3.org/1999/xhtml');
+      foreach ($search_patterns as $search_pattern) {
+        $found = $xml->xpath($search_pattern);
+        if (!empty($found)) {
+          $errors[] = t('Dangerous content found.');
+          break;
+        }
+      }
+      $attributes = $xml->attributes();
+      if (empty($attributes['width'])) {
```

Comment:
Apologies for wasting your time, but after that last suggestion I did a long internet search and apparently it is quite OK for SVGs to be missing dimensions, and in fact some tools remove the dimensions. And since we're displaying only with `img` tags, the dimensions could be set on the image tag anyway. So requiring dimensions is probably overkill?

Reply:
Perhaps we just remove the requirement. Let them expand to the bounds of the containing div if they don't have dimensions.

Reply:
Removed. I played around with it and a nice SVG file can still be viewed fine. If a `viewBox` is present it will still use it to scale to a certain size even if the width/height are missing.

Reply:
Though I had only tested by adding it from `admin/content/file`. If I insert it into a node I get a fatal error: 

> Incorrect integer value: '' for column 'field_vector_image_width'

So either we still need dimensions or we change a db table (and corresponding code).

Reply:
I've had it save zeros if there are no dimensions. But then on image field display we set them to blank so that the image can be visible. But still needs a bit of tweaking with this change.